src/Featurizers/AnalyticalRollingWindowFeaturizer.h

```diff
@@ -0,0 +1,305 @@
+    // |
+    // ----------------------------------------------------------------------
+    AnalyticalRollingWindowTransformer(std::uint32_t windowSize, AnalyticalRollingWindowCalculation windowCalculation, std::uint32_t horizon, std::uint32_t minWindowCount = 1) ;
+    AnalyticalRollingWindowTransformer(Archive &ar);
+    ~AnalyticalRollingWindowTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(AnalyticalRollingWindowTransformer);
+
+    bool operator==(AnalyticalRollingWindowTransformer const &other) const;
+    bool operator!=(AnalyticalRollingWindowTransformer const &other) const;
+    void save(Archive & ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Members
+    // |
+    // ----------------------------------------------------------------------
+    // _windowSizes and _minWindowCount are int64 so that we can have the full positive range of the uint32 it comes
+    // in as but can still support negative numbers.
+    Components::CircularBuffer<InputT>              _buffer;
+    const std::uint32_t                             _windowSize;
+    const std::uint32_t                             _horizon;
+    const std::uint32_t                             _minWindowSize;
+    const AnalyticalRollingWindowCalculation        _windowCalculation;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        typename BaseType::TransformedType         results(_horizon);
+        
+        _buffer.push(input);
+        size_t bufferSize = _buffer.size();
+
+        double result;
+        size_t numElements;
+        size_t startingOffset;
+
+        for (std::uint32_t offset = 0; offset < _horizon; ++offset){
+
+            // If we don't have enough elements then output NaN
+            if ((_horizon - offset) + _minWindowSize > bufferSize) {
+                result = Traits<double>::CreateNullValue();
+
+            // If we have strictly less elements then the window size, but more then the minimum size
+            } else {
+                if (bufferSize - (_horizon - offset) < _windowSize) {
+                    numElements = bufferSize - (_horizon - offset); //asfd
+                    startingOffset = 0;
+                } else {
+                    numElements = _windowSize;
+                    startingOffset = offset - (_buffer.capacity() - bufferSize);
+                }
+
+                const std::tuple<Components::CircularBuffer<InputT>::iterator, Components::CircularBuffer<InputT>::iterator> range = _buffer.range(numElements, startingOffset);

```

Comment:
>const std::tuple<Components::CircularBuffer<InputT>::iterator, Components::CircularBuffer<InputT>::iterator> range = _buffer.range(numElements, startingOffset); [](start = 16, length = 160)

Most (if not all) modern compilers are able to figure this out, but it is similar to a mistake that the compiler isn't smart enough to always figure out....

Taken literally, this assignment will result in the following actions:

1) Default construct range
2) Create a temporary for the result of `_buffer.range`
3) Use range's move assignment operator to move the temporary

This is similar to initializing a member var like this:

`_myFoo(MyFoo(args))`

rather than

`myFoo(args)`

When assigning vars, you should use one of these syntaxes:

A) `range(_buffer.range(_horizon, ...));` [older style used before initializer lists]
B) `range{_buffer.range(_horizon, ...)};` [newer style]

Reply:
Good to know. I have fixed this and the start/end iterators below it too.